src/layers/GeomodelLabelsLayer.ts

```diff
@@ -46,9 +50,40 @@ export class GeomodelLabelsLayer extends CanvasLayer {
   }
 
   setData(data: SurfaceData): void {
+    this.generateSurfaceWithAvgDepth(data);
     super.setData(data);
   }
 
+  generateSurfaceWithAvgDepth(data: any): void {
+    const { areas } = data;
+    this.areasWithAvgTopDepth = areas.reduce((acc: any, a: any) => {
+      // Filter surfaces without label
+      if (!a.label) {
+        return acc;
+      }
+      let topSum = 0;
+      let count = 0;
+      a.data.forEach((d: any) => {
+        if (d[1] != null) {
+          topSum += d[1];
+          count++;
+        }
+      });
```

Comment:
Perhaps use reduce instead?

Reply:
doh

Reply:
Changed

Reply:
A comment regarding this function; the input parameter to the reduce function is mutated making it impure. It is good practice to make reducers pure. Here is a quick example in js of how to make it a little bit more functional and pure:
```
const maybeApply = x => f => o => x ? f(o, x) : o;
const initialObject = { topSum: 0, count: 0, avgTopDepth: null, };
  
const reducer = (a, d) => maybeApply(d[1])((o, x) => ({ 
  topSum: o.topSum + x,
  count: o.count + 1,
  avgTopDepth: (o.topSum + x) / (o.count + 1),
}))(a);

const { avgTopDepth } = area.data.reduce(reducer, initialObject);
```

Reply:
I agree with you in principle, but in this case the data structure of the accumulator is extremely simple and only lives in the reducer except for the single numeric value returned. Also I'm not sure that const maybeApply = x => f => o => x ? f(o, x) : o; and the calling structure is more readable than if (d[1] != null) for whoever is going to maintain this some time in the future. I also believe the test should be x != null as x = 0 would not be counted into the average. 
If there was a strong goal to not mutate the accumulator a simple alternative would be
(a, d) => {
   if (d != null) {
      return{
          topSum: a.topSum + d,
          count: a.count + 1,
          avgTopDepth: (a.topSum + d) / (a.count + 1),
      };
   }
   return a;
};
However this comes with the penalty of creating a new object for every value we iterate over. I tested this in a fiddle and it's about 50% slower than the reducer which reuses the accumulator over a million values. I also tested you suggestion which in the fiddle was roughly twice as slow as the one which reuse the accumulator. I'll add that the timing vary alot but the pattern was clear.

Reply:
Performance wise you are correct of course. Creating a new object for each iteration is slower, especially in a language like Javascript. So if we are dealing with a really large number of iterations, it is not the best solution. In my experience this is rarely the case in frontend apps, so I don´t usually worry about it. But this is not your everyday frontend app I suppose...

Regarding my function replacing the if. The idea is to create a collection of this type of functions and put them somewhere where they are globally available. Then, when you use them a lot, you know what they do and don´t have to look at the code. It will be part of your standard library. Then you can get rid of many if´s and switch-statements which I personally think pollutes the code.

Reply:
In many cases I agree, we should not do premature optimization and in most(all?) cases it is far more important to not mutate input data. But it's also a matter of how much effort you put into optimization and keeping performance in the back of your head. In this case we know we are iterating over some thousands values for each surface, the optimization effort is insignificant and we're sure that what we mutate is not used anywhere else . Given that we have a goal of opening a wellbore or field in <1sec in rep and fetching data takes the better part of that second(if not more), 5ms here and 20ms there quickly add up. (I might add that I stupidly put this code inside the drawing loop at first where it certainly would have an impact.)
I still think the maybeApply  = a => b => c => ... and maybeApply(...)(...)(...) are harder to read and understand, especially for novice programmers which may not have seen such constructs before, even if they are part of an internal lib. An alternative might be something like
const checkValid = d => (d != null);   
    const calcAvg= (a, d) => {
      	a.topSum += d;
      	++a.count;
      	a.avgTopDepth = a.topSum / a.count;
        return a;
    };
const { avgTopDepth } = data.reduce((a, d) => (checkValid(d[1]) ? calcAvg(a, d[1]) : a), {...initialObject});
but even this comes with a performance penalty suggesting that function calls (unsurprisingly?) impact performance 
Another option might be 
const { avgTopDepth } = data.filter(checkValid).reduce(calcAvg, {...initialObject});
however this is three to four(!!!) times slower in my test. (Interestingly, the original forEach loop is slower than the current reducer which is kind of puzzling though access to function parameters might be faster than accessing other variables).
IMHO the four lines of code as is does the job, is satisfactorily easy to read and performant. It'll do. There are better candidates for cleanup and refactoring in the component.